### ISSUE TITLE
How to send ping frame to server or client for websocket?

### ISSUE THREAD

**Issue by gfZeng** [ztellman/aleph#363]
At the server side, How to send `ping frame` to a client?
At the client side, How to send `ping frame` to a server?

Furthermore:
How to determine connection is alive use protocol ping/ pong?

**Comment by kachayev** [comment]
@gfZeng There's no option for doing this right now "out of the box". I planned to submit PR with a support for pings/pongs and idle states processing. 

Somewhat related discussions [here](https://github.com/ztellman/aleph/issues/332) and [here](https://github.com/ztellman/aleph/issues/322).

With the latest version, you can specify a custom `:pipeline-transform` function in order to add a new channel handler to listen to `PingWebSocketFrame` or to send `PongWebSocketFrame` both to client and server connections. It's kinda tricky, but let me play with it and find a solution.

**Comment by kachayev** [comment]
@ztellman That's sounds reasonable 👍 

I have only one problem/question with this. What to do with subsequent pings on a client when I sent `ping` twice before receiving any `pong` from the server? [RFC6455 Section 5.5.3](https://tools.ietf.org/html/rfc6455#section-5.5.3) says

```
  If an endpoint receives a Ping frame and has not yet sent Pong
   frame(s) in response to previous Ping frame(s), the endpoint MAY
   elect to send a Pong frame for only the most recently processed Ping
   frame.
```

Should we prevent users from shouting in their feet? I.e. by resolving deferred to an error right away?

**Comment by ztellman** [comment]
I'd probably just have a PONG trigger *all* pending PINGs, and ignore any subsequent PONGs, since the point of the ping operation is to make sure there's a clear line of communication at that moment, not that a particular ping datagram got a unique response.

**Comment by kachayev** [comment]
Okay! In that case, we can avoid sending all subsequent PINGs at all (as our main goal here is to receive at least one PONG and to resolve all pending waiters).

**Comment by ztellman** [comment]
That works too.

**Comment by gfZeng** [comment]
@kachayev 

I think your version has some bugs on [https://github.com/ztellman/aleph/pull/364/files#diff-649772f881cdedf671330e0bc28a5c56R483](`resolve-pongs`), because of the write & read on the different thread.

Ha, I have written a version of [https://github.com/gfZeng/aleph/commit/a122d4b39b33810a6a05e8fbaa60b97b6a71ca0a](this), my version use `ConcurrentLinkedQueue` instead of atom

**Comment by kachayev** [comment]
@gfZeng Good catch! It was initially done with `swap-vals!` but then I realized it was introduced in 1.9 only, so I removed that and marked PR as WIP 😄to discuss the shape of public API.

I like the idea of passing custom deferred to `ping!`, the only problem is that with your implementation the result of `s/put!` is a deferred that yield `true` when the value is put on the channel, not when `PongFrame` is received. And we still need to resolve deferreds when the netty channel is closed.

**Comment by kachayev** [comment]
@gfZeng Would you mind me talking part of your code (with regards to `ConcurrentLinkedQueue`) to update my implementation? Or maybe to PR into my branch, so I will merge it here in #364? Whatever works.

**Comment by gfZeng** [comment]
@kachayev not at all, help yourself

**Comment by kachayev** [comment]
@gfZeng Thanks! 

@ztellman I updated my implementation, added some basic docs and test cases. I want to add even more tests but probably after per-message compression functionality is merged to avoid overlapping & conflicts.

**Comment by ztellman** [comment]
I'll take a look at this over the weekend.

**Comment by kachayev** [comment]
@ztellman Thanks!

**Comment by kachayev** [comment]
Closed with #364.
